frontend/doc/file-format.rst

```diff
@@ -0,0 +1,324 @@
+ * 4 bytes of chpl major version number
+ * 4 bytes of chpl minor version number
+ * 4 bytes of chpl update version number
+ * 4 bytes of N, the number of module sections contained herein
+ * 256 bits / 32 bytes of SHA-256 hash of the '.dyno' file contents
+   * the hash is computed assuming that the hash is 0
+ * a module section table, consisting of N entries, each consists of:
+   * 8 bytes of module section offset (pointing to a module section header)
+
+Module Section Header
+---------------------
+
+The module section header uses relative section offsets. These relative
+offsets store offsets relative to the module section header.
+
+Each module section begins with a header that consists of:
+
+ * 8 bytes of magic number 0x4d4dd01e5ec14d4d
+ * 8 bytes of reserved space for future flags
+ * 8 bytes module section length
+ * 8 bytes of symbol table section relative offset
+ * 8 bytes of uAST section relative offset
+ * 8 bytes of long strings table section relative offset
+ * 8 bytes of location section relative offset
+ * 8 bytes of types section relative offset
+ * 8 bytes of functions section relative offset
+ * 8 bytes of dependencies section relative offset
+ * a string storing the module symbol ID
+   (e.g. "TopLevelModule" or "MyModule.SubModule")
+   (note that this string does not use the long strings table)
+ * a string storing the file path that generated the module
+   (note that this string does not use the long strings table)
+
+Symbol Table Section
+--------------------
+
+The symbol table section has an entry for each public symbol contained in
+the file. These are sorted by symbol table ID.
+
+For a symbol, the symbol table ID consists of:
+
+ * the name of the symbol, for anything top-level to a module
+ * the regular ID minus the module name, for anything else
+    - Normally, symbol IDs are stored as e.g.  'MyModule.MyClass.myMethod',
+      but since this entire section of the file contains things within
+      'MyModule', that part is omitted; so the symbol table ID for the example
+      would just be 'MyClass.myMethod'.
+
+This section consists of:
+ * 8 bytes of magic number 0x0003bb1e5ec110e0
+ * 4 bytes of N, the number of entries
+ * entries sorted by symbol table ID.  For each entry, it stores:
+   * 8 byte relative offset to the uAST section
+   * 8 byte relative offset to the Locations section
+   * 8 byte relative offset to the type/function entry, if appropriate
+   * a byte storing flags / kind information
+   * a string storing the symbol table ID
+
+uAST Section
+------------
+
+The uAST section consists of serialized uAST entries in the order they
+appear within the file. The top-level entry will be the uAST for the
+corresponding module. Note that the symbol table can point to an
+individual entry within the serialized uAST.
+
+IDs are not stored here. They are recomputed when the uAST is read.
+
+The uAST section consists of:
+
+ * 8 bytes of magic number 0x0003bb1e5ec110e0
```

Comment:
This appears to be the same magic number as the previous section, and would need to be fixed in the code as well.